// @prismicio/helpers/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const types = require('@prismicio/types');
const richtext = require('@prismicio/richtext');
const escapeHtml = require('escape-html');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

const escapeHtml__default = /*#__PURE__*/_interopDefaultLegacy(escapeHtml);

const asDate = (dateOrTimestampField) => {
  if (!dateOrTimestampField) {
    return null;
  }
  if (dateOrTimestampField.length === 24) {
    return new Date(dateOrTimestampField.replace(/(\+|-)(\d{2})(\d{2})$/, ".000$1$2:$3"));
  } else {
    return new Date(dateOrTimestampField);
  }
};

const documentToLinkField = (prismicDocument) => {
  var _a, _b;
  return {
    link_type: types.LinkType.Document,
    id: prismicDocument.id,
    uid: (_a = prismicDocument.uid) != null ? _a : void 0,
    type: prismicDocument.type,
    tags: prismicDocument.tags,
    lang: prismicDocument.lang,
    url: (_b = prismicDocument.url) != null ? _b : void 0,
    slug: prismicDocument.slugs[0],
    ...Object.keys(prismicDocument.data).length > 0 ? { data: prismicDocument.data } : {}
  };
};

const asLink = (linkFieldOrDocument, linkResolver) => {
  if (!linkFieldOrDocument) {
    return null;
  }
  const linkField = "link_type" in linkFieldOrDocument ? linkFieldOrDocument : documentToLinkField(linkFieldOrDocument);
  switch (linkField.link_type) {
    case types.LinkType.Media:
    case types.LinkType.Web:
      return "url" in linkField ? linkField.url : null;
    case types.LinkType.Document:
      if ("url" in linkField && linkField.url) {
        return linkField.url;
      } else if ("id" in linkField) {
        return linkResolver ? linkResolver(linkField) : null;
      } else {
        return null;
      }
    case types.LinkType.Any:
    default:
      return null;
  }
};

const getLabel = (node) => {
  return "data" in node && "label" in node.data ? ` class="${node.data.label}"` : "";
};
const serializeStandardTag = (tag, node, children) => {
  return `<${tag}${getLabel(node)}>${children.join("")}</${tag}>`;
};
const serializePreFormatted = (node) => {
  return `<pre${getLabel(node)}>${escapeHtml__default["default"](node.text)}</pre>`;
};
const serializeImage = (linkResolver, node) => {
  let imageTag = `<img src="${node.url}" alt="${escapeHtml__default["default"](node.alt)}"${node.copyright ? ` copyright="${escapeHtml__default["default"](node.copyright)}"` : ""} />`;
  if (node.linkTo) {
    imageTag = serializeHyperlink(linkResolver, {
      type: types.RichTextNodeType.hyperlink,
      data: node.linkTo,
      start: 0,
      end: 0
    }, [imageTag]);
  }
  return `<p class="block-img">${imageTag}</p>`;
};
const serializeEmbed = (node) => {
  return `<div data-oembed="${node.oembed.embed_url}" data-oembed-type="${node.oembed.type}" data-oembed-provider="${node.oembed.provider_name}"${getLabel(node)}>${node.oembed.html}</div>`;
};
const serializeHyperlink = (linkResolver, node, children) => {
  switch (node.data.link_type) {
    case types.LinkType.Web: {
      return `<a href="${escapeHtml__default["default"](node.data.url)}" target="${node.data.target}" rel="noopener noreferrer"${getLabel(node)}>${children.join("")}</a>`;
    }
    case types.LinkType.Document: {
      return `<a href="${asLink(node.data, linkResolver)}"${getLabel(node)}>${children.join("")}</a>`;
    }
    case types.LinkType.Media: {
      return `<a href="${node.data.url}"${getLabel(node)}>${children.join("")}</a>`;
    }
  }
};
const serializeSpan = (content) => {
  return content ? escapeHtml__default["default"](content).replace(/\n/g, "<br />") : "";
};

function defaultHTMLSerializer(linkResolver, _type, node, content, children, _key) {
  switch (node.type) {
    case richtext.Element.heading1:
      return serializeStandardTag("h1", node, children);
    case richtext.Element.heading2:
      return serializeStandardTag("h2", node, children);
    case richtext.Element.heading3:
      return serializeStandardTag("h3", node, children);
    case richtext.Element.heading4:
      return serializeStandardTag("h4", node, children);
    case richtext.Element.heading5:
      return serializeStandardTag("h5", node, children);
    case richtext.Element.heading6:
      return serializeStandardTag("h6", node, children);
    case richtext.Element.paragraph:
      return serializeStandardTag("p", node, children);
    case richtext.Element.preformatted:
      return serializePreFormatted(node);
    case richtext.Element.strong:
      return serializeStandardTag("strong", node, children);
    case richtext.Element.em:
      return serializeStandardTag("em", node, children);
    case richtext.Element.listItem:
      return serializeStandardTag("li", node, children);
    case richtext.Element.oListItem:
      return serializeStandardTag("li", node, children);
    case richtext.Element.list:
      return serializeStandardTag("ul", node, children);
    case richtext.Element.oList:
      return serializeStandardTag("ol", node, children);
    case richtext.Element.image:
      return serializeImage(linkResolver, node);
    case richtext.Element.embed:
      return serializeEmbed(node);
    case richtext.Element.hyperlink:
      return serializeHyperlink(linkResolver, node, children);
    case richtext.Element.label:
      return serializeStandardTag("span", node, children);
    case richtext.Element.span:
    default:
      return serializeSpan(content);
  }
}
function asHTML(richTextField, linkResolver, htmlSerializer) {
  let serializer;
  if (htmlSerializer) {
    serializer = richtext.composeSerializers(typeof htmlSerializer === "object" ? richtext.wrapMapSerializer(htmlSerializer) : htmlSerializer, defaultHTMLSerializer.bind(null, linkResolver));
  } else {
    serializer = defaultHTMLSerializer.bind(null, linkResolver);
  }
  return richtext.serialize(richTextField, serializer).join("");
}

Object.defineProperty(exports, 'Element', {
	enumerable: true,
	get: function () { return richtext.Element; }
});
Object.defineProperty(exports, 'asText', {
	enumerable: true,
	get: function () { return richtext.asText; }
});
exports.asDate = asDate;
exports.asHTML = asHTML;
exports.asLink = asLink;
exports.documentToLinkField = documentToLinkField;
