// fp-ts-contrib/lib/ArrayOption.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Array_1 = require("fp-ts/lib/Array");
var function_1 = require("fp-ts/lib/function");
var OptionT_1 = require("fp-ts/lib/OptionT");
var pipeable_1 = require("fp-ts/lib/pipeable");
var T = OptionT_1.getOptionM(Array_1.array);
// -------------------------------------------------------------------------------------
// constructors
// -------------------------------------------------------------------------------------
/**
 * @category constructors
 * @since 0.1.0
 */
exports.fromArray = T.fromM;
/**
 * @category constructors
 * @since 0.1.0
 */
exports.fromOption = Array_1.of;
/**
 * @category constructors
 * @since 0.1.10
 */
exports.fromOptionK = function (f) { return function () {
    var a = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        a[_i] = arguments[_i];
    }
    return exports.fromOption(f.apply(void 0, a));
}; };
/**
 * @category constructors
 * @since 0.1.0
 */
exports.none = T.none();
/**
 * @category constructors
 * @since 0.1.0
 */
exports.some = T.of;
// -------------------------------------------------------------------------------------
// destructors
// -------------------------------------------------------------------------------------
/**
 * @category destructors
 * @since 0.1.0
 */
exports.fold = function (onNone, onSome) { return function (as) { return T.fold(as, onNone, onSome); }; };
/**
 * @category destructors
 * @since 0.1.0
 */
exports.getOrElse = function (onNone) { return function (as) {
    return T.getOrElse(as, onNone);
}; };
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @category Functor
 * @since 0.1.18
 */
exports.map = function (f) { return function (fa) { return T.map(fa, f); }; };
/**
 * @category Apply
 * @since 0.1.18
 */
exports.ap = function (fa) { return function (fab) {
    return T.ap(fab, fa);
}; };
/**
 * @category Apply
 * @since 0.1.18
 */
exports.apFirst = function (fb) { return function (fa) {
    return pipeable_1.pipe(fa, exports.map(function (a) { return function (_) { return a; }; }), exports.ap(fb));
}; };
/**
 * @category Apply
 * @since 0.1.18
 */
exports.apSecond = function (fb) { return function (fa) {
    return pipeable_1.pipe(fa, exports.map(function () { return function (b) { return b; }; }), exports.ap(fb));
}; };
/**
 * @category Applicative
 * @since 0.1.18
 */
exports.of = T.of;
/**
 * @category Monad
 * @since 0.1.10
 */
exports.chainOptionK = function (f) {
    return exports.chain(exports.fromOptionK(f));
};
/**
 * @category Monad
 * @since 0.1.18
 */
exports.chain = function (f) { return function (fa) {
    return T.chain(fa, f);
}; };
/**
 * @category Monad
 * @since 0.1.18
 */
exports.chainFirst = function (f) { return function (ma) {
    return T.chain(ma, function (a) { return T.map(f(a), function () { return a; }); });
}; };
/**
 * @category Monad
 * @since 0.1.18
 */
exports.flatten = function (mma) { return T.chain(mma, function_1.identity); };
/**
 * @category Alternative
 * @since 0.1.18
 */
exports.alt = function (that) { return function (fa) {
    return T.alt(fa, that);
}; };
/**
 * @category Alternative
 * @since 0.1.18
 */
exports.zero = function () { return Array_1.empty; };
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @category instances
 * @since 0.1.0
 */
exports.URI = 'ArrayOption';
/**
 * @category instances
 * @since 0.1.18
 */
exports.Functor = {
    URI: exports.URI,
    map: T.map
};
/**
 * @category instances
 * @since 0.1.18
 */
exports.Applicative = {
    URI: exports.URI,
    map: T.map,
    ap: T.ap,
    of: exports.of
};
/**
 * @category instances
 * @since 0.1.18
 */
exports.Apply = {
    URI: exports.URI,
    ap: T.ap,
    map: T.map
};
/**
 * @category instances
 * @since 0.1.18
 */
exports.Monad = {
    URI: exports.URI,
    ap: T.ap,
    chain: T.chain,
    map: T.map,
    of: T.of
};
/**
 * @category instances
 * @since 0.1.18
 */
exports.Alt = {
    URI: exports.URI,
    alt: T.alt,
    map: T.map
};
/**
 * @category instances
 * @since 0.1.18
 */
exports.Alternative = {
    URI: exports.URI,
    alt: T.alt,
    ap: T.ap,
    map: T.map,
    of: T.of,
    zero: exports.zero
};
/**
 * @category instances
 * @since 0.1.0
 */
exports.arrayOption = {
    URI: exports.URI,
    map: T.map,
    of: exports.some,
    ap: T.ap,
    chain: T.chain,
    alt: T.alt
};
