// fp-ts-contrib/lib/Free.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var function_1 = require("fp-ts/lib/function");
// -------------------------------------------------------------------------------------
// constructors
// -------------------------------------------------------------------------------------
var impure = function (fx, f) { return ({ _tag: 'Impure', fx: fx, f: f }); };
/**
 * Lift an impure value described by the generating type constructor `F` into the free monad
 *
 * @category constructors
 * @since 0.1.3
 */
exports.liftF = function (fa) { return impure(fa, function (a) { return exports.free.of(a); }); };
function foldFree(M) {
    return function (nt, fa) {
        if (exports.isPure(fa)) {
            return M.of(fa.value);
        }
        else {
            return M.chain(nt(fa.fx), function (x) { return foldFree(M)(nt, fa.f(x)); });
        }
    };
}
exports.foldFree = foldFree;
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
var substFree = function (f) {
    function go(fa) {
        switch (fa._tag) {
            case 'Pure':
                return exports.free.of(fa.value);
            case 'Impure':
                return exports.free.chain(f(fa.fx), function (x) { return go(fa.f(x)); });
        }
    }
    return go;
};
function hoistFree(nt) {
    return substFree(function (fa) { return exports.liftF(nt(fa)); });
}
exports.hoistFree = hoistFree;
// -------------------------------------------------------------------------------------
// non-pipeables
// -------------------------------------------------------------------------------------
var map_ = function (fa, f) {
    return exports.isPure(fa) ? exports.free.of(f(fa.value)) : impure(fa.fx, function (x) { return exports.free.map(fa.f(x), f); });
};
var ap_ = function (fab, fa) {
    return exports.free.chain(fab, function (f) { return exports.free.map(fa, f); });
};
var chain_ = function (ma, f) {
    return exports.isPure(ma) ? f(ma.value) : impure(ma.fx, function (x) { return exports.free.chain(ma.f(x), f); });
};
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @category Functor
 * @since 0.1.18
 */
exports.map = function (f) { return function (fa) { return map_(fa, f); }; };
/**
 * @category Apply
 * @since 0.1.18
 */
exports.ap = function (fa) { return function (fab) { return ap_(fab, fa); }; };
/**
 * @category Monad
 * @since 0.1.18
 */
exports.chain = function (f) { return function (ma) { return chain_(ma, f); }; };
/**
 * @category Monad
 * @since 0.1.18
 */
exports.flatten = function (mma) { return chain_(mma, function_1.identity); };
/**
 * @category Applicative
 * @since 0.1.18
 */
exports.of = function (a) { return ({ _tag: 'Pure', value: a }); };
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @category instances
 * @since 0.1.3
 */
exports.URI = 'Free';
/**
 * @category instances
 * @since 0.1.18
 */
exports.Functor = {
    URI: exports.URI,
    map: map_
};
/**
 * @category instances
 * @since 0.1.18
 */
exports.Applicative = {
    URI: exports.URI,
    map: map_,
    ap: ap_,
    of: exports.of
};
/**
 * @category instances
 * @since 0.1.18
 */
exports.Apply = {
    URI: exports.URI,
    ap: ap_,
    map: map_
};
/**
 * Monad instance for Free
 *
 * @category instances
 * @since 0.1.3
 */
exports.free = {
    URI: exports.URI,
    ap: ap_,
    chain: chain_,
    map: map_,
    of: exports.of
};
// -------------------------------------------------------------------------------------
// utils
// -------------------------------------------------------------------------------------
/**
 * Check if given Free instance is Pure
 *
 * @since 0.1.3
 */
exports.isPure = function (fa) { return fa._tag === 'Pure'; };
/**
 * Check if given Free instance is Impure
 *
 * @since 0.1.3
 */
exports.isImpure = function (fa) { return fa._tag === 'Impure'; };
