// fp-ts-contrib/lib/List.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var A = __importStar(require("fp-ts/lib/Array"));
var Foldable_1 = require("fp-ts/lib/Foldable");
var function_1 = require("fp-ts/lib/function");
var Monoid_1 = require("fp-ts/lib/Monoid");
var O = __importStar(require("fp-ts/lib/Option"));
var pipeable_1 = require("fp-ts/lib/pipeable");
// -------------------------------------------------------------------------------------
// constructors
// -------------------------------------------------------------------------------------
/**
 * @category constructors
 * @since 0.1.8
 */
exports.nil = { type: 'Nil', length: 0 };
/**
 * Attaches an element to the front of a list.
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.deepStrictEqual(L.cons('a', L.nil), { type: 'Cons', head: 'a', tail: L.nil, length: 1 })
 *
 * @category constructors
 * @since 0.1.8
 */
exports.cons = function (head, tail) { return ({
    type: 'Cons',
    head: head,
    tail: tail,
    length: 1 + tail.length
}); };
/**
 * Creates a list from an array
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.deepStrictEqual(L.fromArray([]), L.nil)
 * assert.deepStrictEqual(L.fromArray(['a', 'b']), L.cons('a', L.of('b')))
 *
 * @category constructors
 * @since 0.1.8
 */
exports.fromArray = function (as) { return A.array.reduceRight(as, exports.nil, exports.cons); };
// -------------------------------------------------------------------------------------
// destructors
// -------------------------------------------------------------------------------------
/**
 * Gets the first element in a list, or `None` if the list is empty.
 *
 * @example
 * import * as O from 'fp-ts/Option'
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.deepStrictEqual(L.head(L.nil), O.none)
 * assert.deepStrictEqual(L.head(L.cons('x', L.of('a'))), O.some('x'))
 *
 * @category destructors
 * @since 0.1.8
 */
exports.head = function (fa) { return (exports.isCons(fa) ? O.some(fa.head) : O.none); };
/**
 * Gets all but the first element of a list, or `None` if the list is empty.
 *
 * @example
 * import * as O from 'fp-ts/Option'
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.deepStrictEqual(L.tail(L.nil), O.none)
 * assert.deepStrictEqual(L.tail(L.of('a')), O.some(L.nil))
 * assert.deepStrictEqual(L.tail(L.cons('x', L.of('a'))), O.some(L.of('a')))
 *
 * @category destructors
 * @since 0.1.8
 */
exports.tail = function (fa) { return (exports.isCons(fa) ? O.some(fa.tail) : O.none); };
/**
 * Breaks a list into its first element and the remaining elements.
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * const len: <A>(as: L.List<A>) => number = L.foldLeft(
 *   () => 0,
 *   (_, tail) => 1 + len(tail)
 * )
 * assert.deepStrictEqual(len(L.cons('a', L.of('b'))), 2)
 *
 * @category destructors
 * @since 0.1.8
 */
exports.foldLeft = function (onNil, onCons) { return function (fa) { return (exports.isNil(fa) ? onNil() : onCons(fa.head, fa.tail)); }; };
/**
 * Gets an array from a list.
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.deepStrictEqual(L.toArray(L.cons('a', L.of('b'))), ['a', 'b'])
 *
 * @category destructors
 * @since 0.1.8
 */
exports.toArray = function (fa) {
    var length = fa.length;
    var out = new Array(length);
    var l = fa;
    for (var i = 0; i < length; i++) {
        out[i] = l.head;
        l = l.tail;
    }
    return out;
};
/**
 * Gets an array from a list in a reversed order.
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.deepStrictEqual(L.toReversedArray(L.cons('a', L.of('b'))), ['b', 'a'])
 *
 * @category destructors
 * @since 0.1.8
 */
exports.toReversedArray = function (fa) {
    var length = fa.length;
    var out = new Array(length);
    var l = fa;
    for (var i = 0; i < length; i++) {
        out[length - i - 1] = l.head;
        l = l.tail;
    }
    return out;
};
// -------------------------------------------------------------------------------------
// combinators
// -------------------------------------------------------------------------------------
/**
 * Reverse a list.
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.deepStrictEqual(L.reverse(L.cons(1, L.cons(2, L.of(3)))), L.cons(3, L.cons(2, L.of(1))))
 *
 * @category combinators
 * @since 0.1.8
 */
exports.reverse = function (fa) {
    var out = exports.nil;
    var l = fa;
    while (exports.isCons(l)) {
        out = exports.cons(l.head, out);
        l = l.tail;
    }
    return out;
};
/**
 * Drops the specified number of elements from the front of a list.
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.deepStrictEqual(L.dropLeft(1)(L.nil), L.nil)
 * assert.deepStrictEqual(L.dropLeft(1)(L.cons(1, L.of(2))), L.of(2))
 * assert.deepStrictEqual(L.dropLeft(3)(L.cons(1, L.of(2))), L.nil)
 *
 * @category combinators
 * @since 0.1.8
 */
exports.dropLeft = function (n) { return function (fa) {
    if (exports.isNil(fa))
        return exports.nil;
    var i = 0;
    var l = fa;
    while (exports.isCons(l) && i < n) {
        i++;
        l = l.tail;
    }
    return l;
}; };
function dropLeftWhile(predicate) {
    return function (fa) {
        if (exports.isNil(fa))
            return exports.nil;
        var l = fa;
        while (exports.isCons(l) && predicate(l.head)) {
            l = l.tail;
        }
        return l;
    };
}
exports.dropLeftWhile = dropLeftWhile;
// -------------------------------------------------------------------------------------
// non-pipeables
// -------------------------------------------------------------------------------------
var map_ = function (fa, f) { return pipeable_1.pipe(fa, exports.map(f)); };
var ap_ = function (fab, fa) { return pipeable_1.pipe(fab, exports.ap(fa)); };
var chain_ = function (ma, f) { return pipeable_1.pipe(ma, exports.chain(f)); };
var reduce_ = function (fa, b, f) { return pipeable_1.pipe(fa, exports.reduce(b, f)); };
var reduceRight_ = function (fa, b, f) { return pipeable_1.pipe(fa, exports.reduceRight(b, f)); };
var foldMap_ = function (M) { return function (fa, f) { return pipeable_1.pipe(fa, exports.foldMap(M)(f)); }; };
var traverse_ = function (F) {
    return function (ta, f) {
        return exports.list.reduceRight(ta, F.of(exports.nil), function (a, fbs) {
            return F.ap(F.map(fbs, function (bs) { return function (b) { return exports.cons(b, bs); }; }), f(a));
        });
    };
};
var sequence_ = function (F) { return function (ta) {
    return exports.list.reduceRight(ta, F.of(exports.nil), function (a, fas) {
        return F.ap(F.map(fas, function (as) { return function (a) { return exports.cons(a, as); }; }), a);
    });
}; };
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @category Functor
 * @since 0.1.18
 */
exports.map = function (f) { return function (fa) {
    return pipeable_1.pipe(exports.toArray(fa), A.reduceRight(exports.nil, function (a, b) { return exports.cons(f(a), b); }));
}; };
/**
 * @category Functor
 * @since 0.1.20
 */
exports.ap = function (fa) { return exports.chain(function (f) { return pipeable_1.pipe(fa, exports.map(f)); }); };
/**
 * @category Apply
 * @since 0.1.20
 */
exports.apFirst = function (fb) {
    return function_1.flow(exports.map(function (a) { return function () { return a; }; }), exports.ap(fb));
};
/**
 * @category Apply
 * @since 0.1.20
 */
exports.apSecond = function (fb) {
    return function_1.flow(exports.map(function () { return function (b) { return b; }; }), exports.ap(fb));
};
/**
 * @category Monad
 * @since 0.1.20
 */
exports.chain = function (f) { return function (ma) {
    return pipeable_1.pipe(ma, exports.foldLeft(function () { return exports.nil; }, function (x, xs) {
        var S = exports.getSemigroup();
        var out = f(x);
        var l = xs;
        while (exports.isCons(l)) {
            out = S.concat(out, f(l.head));
            l = l.tail;
        }
        return out;
    }));
}; };
/**
 * @category Monad
 * @since 0.1.20
 */
exports.chainFirst = function (f) {
    return exports.chain(function (a) {
        return pipeable_1.pipe(f(a), exports.map(function () { return a; }));
    });
};
/**
 * @category Foldable
 * @since 0.1.18
 */
exports.reduce = function (b, f) { return function (fa) {
    var out = b;
    var l = fa;
    while (exports.isCons(l)) {
        out = f(out, l.head);
        l = l.tail;
    }
    return out;
}; };
/**
 * @category Foldable
 * @since 0.1.18
 */
exports.reduceRight = function (b, f) { return function (fa) {
    return pipeable_1.pipe(exports.toArray(fa), A.reduceRight(b, f));
}; };
/**
 * @category Foldable
 * @since 0.1.18
 */
exports.foldMap = function (M) { return function (f) { return function (fa) {
    var out = M.empty;
    var l = fa;
    while (exports.isCons(l)) {
        out = M.concat(out, f(l.head));
        l = l.tail;
    }
    return out;
}; }; };
// TODO: add pipeable traverse when fp-ts version >= 2.6.3
// /**
//  * @category Traversable
//  * @since 0.1.18
//  */
// export const traverse: PipeableTraverse1<URI> = <F>(
//   F: Applicative<F>
// ): (<A, B>(f: (a: A) => HKT<F, B>) => (ta: List<A>) => HKT<F, List<B>>) => {
//   const traverseF = traverse_(F)
//   return f => ta => traverseF(ta, f)
// }
/**
 * @category Traversable
 * @since 0.1.18
 */
exports.sequence = function (F) { return sequence_(F); };
/**
 * Creates a list with a single element.
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.deepStrictEqual(L.of('a'), L.cons('a', L.nil))
 *
 * @category Applicative
 * @since 0.1.8
 */
exports.of = function (head) { return exports.cons(head, exports.nil); };
// -------------------------------------------------------------------------------------
// utils
// -------------------------------------------------------------------------------------
/**
 * Finds the first index for which a predicate holds.
 *
 * @example
 * import * as O from 'fp-ts/Option'
 * import * as L from 'fp-ts-contrib/List'
 *
 * const f = (a: number): boolean => a % 2 === 0
 * const findIndexEven = L.findIndex(f)
 * assert.deepStrictEqual(findIndexEven(L.nil), O.none)
 * assert.deepStrictEqual(findIndexEven(L.cons(1, L.of(2))), O.some(1))
 * assert.deepStrictEqual(findIndexEven(L.of(1)), O.none)
 *
 * @since 0.1.8
 */
exports.findIndex = function (predicate) { return function (fa) {
    var l = fa;
    var i = 0;
    while (exports.isCons(l)) {
        if (predicate(l.head))
            return O.some(i);
        l = l.tail;
        i++;
    }
    return O.none;
}; };
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @category instances
 * @since 0.1.8
 */
exports.URI = 'List';
/**
 * Derives an `Eq` over the `List` of a given element type from the `Eq` of that type.
 * The derived `Eq` defines two lists as equal if all elements of both lists
 * are compared equal pairwise with the given `E`. In case of lists of different
 * lengths, the result is non equality.
 *
 * @example
 * import { eqString } from 'fp-ts/Eq'
 * import * as L from 'fp-ts-contrib/List'
 *
 * const E = L.getEq(eqString)
 * assert.strictEqual(E.equals(L.cons('a', L.of('b')), L.cons('a', L.of('b'))), true)
 * assert.strictEqual(E.equals(L.of('x'), L.nil), false)
 *
 * @category instances
 * @since 0.1.8
 */
exports.getEq = function (E) { return ({
    equals: function (x, y) {
        if (x.length !== y.length)
            return false;
        var lx = x;
        var ly = y;
        while (exports.isCons(lx) && exports.isCons(ly)) {
            if (!E.equals(lx.head, ly.head))
                return false;
            lx = lx.tail;
            ly = ly.tail;
        }
        return true;
    }
}); };
/**
 * @category instances
 * @since 0.1.20
 */
exports.getShow = function (S) { return ({
    show: function (as) {
        return pipeable_1.pipe(as, exports.map(S.show), exports.foldLeft(function () { return 'Nil'; }, function (x, xs) { return "(" + Foldable_1.intercalate(Monoid_1.monoidString, exports.Foldable)(' : ', exports.cons(x, xs)) + " : Nil)"; }));
    }
}); };
/**
 * @category instances
 * @since 0.1.20
 */
exports.getSemigroup = function () { return ({
    concat: function (xs, ys) { return pipeable_1.pipe(xs, exports.reduceRight(ys, exports.cons)); }
}); };
/**
 * @category instances
 * @since 0.1.20
 */
exports.getMonoid = function () { return (__assign(__assign({}, exports.getSemigroup()), { empty: exports.nil })); };
/**
 * @category instances
 * @since 0.1.18
 */
exports.Functor = {
    URI: exports.URI,
    map: map_
};
/**
 * @category instances
 * @since 0.1.20
 */
exports.Apply = {
    URI: exports.URI,
    map: map_,
    ap: ap_
};
/**
 * @category instances
 * @since 0.1.20
 */
exports.Applicative = {
    URI: exports.URI,
    map: map_,
    ap: ap_,
    of: exports.of
};
/**
 * @category instances
 * @since 0.1.20
 */
exports.Monad = {
    URI: exports.URI,
    map: map_,
    ap: ap_,
    of: exports.of,
    chain: chain_
};
/**
 * @category instances
 * @since 0.1.18
 */
exports.Foldable = {
    URI: exports.URI,
    foldMap: foldMap_,
    reduce: reduce_,
    reduceRight: reduceRight_
};
/**
 * @category instances
 * @since 0.1.18
 */
exports.Traversable = {
    URI: exports.URI,
    map: map_,
    foldMap: foldMap_,
    reduce: reduce_,
    reduceRight: reduceRight_,
    traverse: traverse_,
    sequence: sequence_
};
/**
 * @category instances
 * @since 0.1.8
 */
exports.list = {
    URI: exports.URI,
    map: map_,
    reduce: reduce_,
    foldMap: foldMap_,
    reduceRight: reduceRight_,
    traverse: traverse_,
    sequence: sequence_
};
// -------------------------------------------------------------------------------------
// do notation
// -------------------------------------------------------------------------------------
/**
 * @internal
 */
var bind_ = function (a, name, b) {
    var _a;
    return Object.assign({}, a, (_a = {}, _a[name] = b, _a));
};
/**
 * @since 0.1.20
 */
exports.bindTo = function (name) { return function (fa) {
    return pipeable_1.pipe(fa, exports.map(function (a) { return bind_({}, name, a); }));
}; };
/**
 * @since 0.1.20
 */
exports.bind = function (name, f) { return function (fa) {
    return pipeable_1.pipe(fa, exports.chain(function (a) {
        return pipeable_1.pipe(f(a), exports.map(function (b) { return bind_(a, name, b); }));
    }));
}; };
// -------------------------------------------------------------------------------------
// pipeable sequence S
// -------------------------------------------------------------------------------------
/**
 * @since 0.1.20
 */
exports.apS = function (name, fb) {
    return function_1.flow(exports.map(function (a) { return function (b) { return bind_(a, name, b); }; }), exports.ap(fb));
};
// -------------------------------------------------------------------------------------
// utils
// -------------------------------------------------------------------------------------
/**
 * Tests whether a list is an empty list.
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.strictEqual(L.isNil(L.nil), true)
 * assert.strictEqual(L.isNil(L.of(6)), false)
 *
 * @since 0.1.8
 */
exports.isNil = function (a) { return a.type === 'Nil'; };
/**
 * Tests whether a list is a non empty list.
 *
 * @example
 * import * as L from 'fp-ts-contrib/List'
 *
 * assert.strictEqual(L.isCons(L.nil), false)
 * assert.strictEqual(L.isCons(L.of(1)), true)
 *
 * @since 0.1.8
 */
exports.isCons = function (a) { return a.type === 'Cons'; };
